File: build_xlsx.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { SpreadsheetFile, Workbook } from "@oai/artifact-tool";

const [jsonPath, outputPath] = process.argv.slice(2);
if (!jsonPath || !outputPath) throw new Error("缺少輸入或輸出路徑。");
const { headers, rows } = JSON.parse(await fs.readFile(jsonPath, "utf8"));
const required = ["路段", "交叉路名", "發生時間", "肇事原因", "年齡", "當事者區分", "件數"];
const lastRow = rows.length + 1;
const missing = required.filter(name => !headers.includes(name));
if (missing.length) throw new Error(`Pivot Cache 缺少必要欄位：${missing.join("、")}`);

const toColumn = n => { let s = ""; for (n += 1; n; n = Math.floor((n - 1) / 26)) s = String.fromCharCode(65 + ((n - 1) % 26)) + s; return s; };
const field = Object.fromEntries(headers.map((h, i) => [h, toColumn(i)]));
const index = Object.fromEntries(headers.map((h, i) => [h, i]));
const count = row => Number(row[index["件數"]] || 0);
function topGroups(valueFn, max = 20) {
  const values = new Map();
  for (const row of rows) {
    const value = valueFn(row);
    if (value !== null && value !== undefined && String(value).trim() !== "") values.set(String(value), (values.get(String(value)) || 0) + count(row));
  }
  return [...values.entries()].sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0], "zh-Hant")).slice(0, max).map(([value]) => value);
}
const roads = topGroups(r => r[index["路段"]]);
const intersections = topGroups(r => r[index["路段"]] && r[index["交叉路名"]] ? `${r[index["路段"]]}／${r[index["交叉路名"]]}` : null);
const periods = topGroups(r => r[index["發生時間"]], 24);
const causes = topGroups(r => r[index["肇事原因"]]);
const vehicles = topGroups(r => r[index["當事者區分"]]);

const wb = Workbook.create();
const source = wb.worksheets.add("原始資料");
source.showGridLines = false;
source.getRange("A1:BX1").values = [[...headers, "路口"]];
source.getRange(`A2:${toColumn(headers.length - 1)}${rows.length + 1}`).values = rows;
source.getRange("BX2").formulas = [[`=IF(AND($${field["路段"]}2<>\"\",$${field["交叉路名"]}2<>\"\"),$${field["路段"]}2&\"／\"&$${field["交叉路名"]}2,\"\")`]];
source.getRange(`BX2:BX${lastRow}`).fillDown();
const table = source.tables.add(`A1:BX${rows.length + 1}`, true, "RecoveredPivotData");
table.style = "TableStyleMedium2";
table.showFilterButton = true;
source.getRange("A1:BX1").format = { fill: "#1F4E78", font: { bold: true, color: "#FFFFFF" } };
source.getRange("A:BX").format.columnWidth = 15;
source.getRange(`${field["路段"]}:${field["路段"]}`).format.columnWidth = 22;
source.getRange(`${field["肇事原因"]}:${field["肇事原因"]}`).format.columnWidth = 34;
source.freezePanes.freezeRows(1);

const sheet = wb.worksheets.add("分析表");
sheet.showGridLines = false;
sheet.mergeCells("A1:H1");
sheet.getRange("A1").values = [["交通事故統計分析"]];
sheet.getRange("A1:H1").format = { fill: "#1F4E78", font: { bold: true, color: "#FFFFFF", size: 14 }, horizontalAlignment: "center" };
sheet.getRange("A1:H1").format.rowHeight = 28;
sheet.mergeCells("A2:H2");
sheet.getRange("A2").values = [["本表由樞紐快取自動還原；藍底數量為公式加總。"]];
sheet.getRange("A2:H2").format = { font: { italic: true, color: "#4B5563", size: 10 } };

function header(left, right, row, title) {
  sheet.getRange(`${left}${row}:${right}${row}`).values = [[title, "數量"]];
  sheet.getRange(`${left}${row}:${right}${row}`).format = { fill: "#D9EAF7", font: { bold: true } };
  sheet.getRange(`${left}${row + 1}`).values = [["依輸入資料自動加總"]];
  sheet.getRange(`${left}${row + 1}`).format = { font: { italic: true, color: "#4B5563", size: 10 } };
}
function section(left, right, title, labels, sourceColumn, startRow, totalRow) {
  header(left, right, startRow - 2, title);
  sheet.getRange(`${left}${startRow}:${left}${startRow + labels.length - 1}`).values = labels.map(v => [v]);
  sheet.getRange(`${right}${startRow}:${right}${startRow + labels.length - 1}`).formulas = labels.map((_, i) => [`=SUMIF('原始資料'!$${sourceColumn}$2:$${sourceColumn}$${lastRow},${left}${startRow + i},'原始資料'!$${field["件數"]}$2:$${field["件數"]}$${lastRow})`]);
  sheet.getRange(`${left}${startRow}:${right}${startRow + labels.length - 1}`).format.borders = { preset: "insideHorizontal", style: "thin", color: "#D9E2F3" };
  sheet.getRange(`${right}${startRow}:${right}${startRow + labels.length - 1}`).format = { fill: "#DDEBF7", numberFormat: "#,##0" };
  sheet.getRange(`${left}${totalRow}`).values = [[`${title}合計`]];
  sheet.getRange(`${right}${totalRow}`).formulas = [[`=SUM('原始資料'!$${field["件數"]}$2:$${field["件數"]}$${lastRow})`]];
  sheet.getRange(`${left}${totalRow}:${right}${totalRow}`).format = { fill: "#D9EAF7", font: { bold: true }, numberFormat: "#,##0", borders: { preset: "doubleBottom", style: "thin", color: "#7F8C8D" } };
}

section("A", "B", "易肇事路段", roads, field["路段"], 6, 27);
header("D", "E", 4, "易肇事路口");
sheet.getRange("D6:D25").values = intersections.map(v => [v]);
sheet.getRange("E6:E25").formulas = intersections.map((_, i) => [`=SUMIF('原始資料'!$BX$2:$BX$${lastRow},D${6 + i},'原始資料'!$${field["件數"]}$2:$${field["件數"]}$${lastRow})`]);
sheet.getRange("D6:E25").format.borders = { preset: "insideHorizontal", style: "thin", color: "#D9E2F3" };
sheet.getRange("E6:E25").format = { fill: "#DDEBF7", numberFormat: "#,##0" };
sheet.getRange("D27:E27").values = [["易肇事路口合計", null]];
sheet.getRange("E27").formulas = [[`=COUNTIFS('原始資料'!$${field["路段"]}$2:$${field["路段"]}$${lastRow},\"<>\",'原始資料'!$${field["交叉路名"]}$2:$${field["交叉路名"]}$${lastRow},\"<>\")`]];
sheet.getRange("D27:E27").format = { fill: "#D9EAF7", font: { bold: true }, numberFormat: "#,##0", borders: { preset: "doubleBottom", style: "thin", color: "#7F8C8D" } };
section("G", "H", "易肇事時段", periods, field["發生時間"], 6, 31);
section("A", "B", "易肇事肇因", causes, field["肇事原因"], 43, 64);

header("D", "E", 41, "易肇事年齡");
sheet.getRange("D42:D51").values = [["0-17"],["18-20"],["21-30"],["31-40"],["41-50"],["51-60"],["61-70"],["71歲以上"],["沒有年紀"],["合計"]];
const intervals = [[0,17],[18,20],[21,30],[31,40],[41,50],[51,60],[61,70]];
const ageFx = intervals.map(([lo, hi]) => `=SUMIFS('原始資料'!$${field["件數"]}$2:$${field["件數"]}$${lastRow},'原始資料'!$${field["年齡"]}$2:$${field["年齡"]}$${lastRow},\">=${lo}\",'原始資料'!$${field["年齡"]}$2:$${field["年齡"]}$${lastRow},\"<=${hi}\",'原始資料'!$${field["年齡"]}$2:$${field["年齡"]}$${lastRow},\"<>\")`);
ageFx.push(`=SUMIFS('原始資料'!$${field["件數"]}$2:$${field["件數"]}$${lastRow},'原始資料'!$${field["年齡"]}$2:$${field["年齡"]}$${lastRow},\">=71\",'原始資料'!$${field["年齡"]}$2:$${field["年齡"]}$${lastRow},\"<>\")`);
ageFx.push(`=SUM('原始資料'!$${field["件數"]}$2:$${field["件數"]}$${lastRow})-SUM(E42:E49)`);
ageFx.push("=SUM(E42:E50)");
sheet.getRange("E42:E51").formulas = ageFx.map(formula => [formula]);
sheet.getRange("D42:E50").format.borders = { preset: "insideHorizontal", style: "thin", color: "#D9E2F3" };
sheet.getRange("E42:E51").format = { fill: "#DDEBF7", numberFormat: "#,##0" };
sheet.getRange("D51:E51").format = { fill: "#D9EAF7", font: { bold: true }, numberFormat: "#,##0", borders: { preset: "doubleBottom", style: "thin", color: "#7F8C8D" } };
section("G", "H", "易肇事車種", vehicles, field["當事者區分"], 43, 64);

for (const [column, width] of [["A",28],["B",13],["C",4],["D",32],["E",13],["F",4],["G",34],["H",13]]) sheet.getRange(`${column}:${column}`).format.columnWidth = width;
sheet.getRange("A1:H64").format.wrapText = false;
sheet.freezePanes.freezeRows(4);

await fs.mkdir(path.dirname(outputPath), { recursive: true });
const out = await SpreadsheetFile.exportXlsx(wb);
await out.save(outputPath);

File: build_xlsx.py
#!/usr/bin/env python3
"""Create the standalone accident-analysis workbook with public Python libraries."""

from __future__ import annotations

import json
import sys
from collections import defaultdict
from pathlib import Path

from openpyxl import Workbook
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side
from openpyxl.utils import get_column_letter
from openpyxl.worksheet.table import Table, TableStyleInfo


REQUIRED = ("路段", "交叉路名", "發生時間", "肇事原因", "年齡", "當事者區分", "件數")
BLUE = "1F4E78"
LIGHT_BLUE = "D9EAF7"
VALUE_BLUE = "DDEBF7"
THIN = Side(style="thin", color="D9E2F3")


def _rank(rows, index, count_index, maximum=20):
    totals = defaultdict(float)
    for row in rows:
        value = str(row[index] or "").strip()
        if value:
            totals[value] += float(row[count_index] or 0)
    return [key for key, _ in sorted(totals.items(), key=lambda item: (-item[1], item[0]))[:maximum]]


def _intersection_rank(rows, road_index, cross_index, count_index):
    totals = defaultdict(float)
    for row in rows:
        road, cross = str(row[road_index] or "").strip(), str(row[cross_index] or "").strip()
        if road and cross:
            totals[f"{road}／{cross}"] += float(row[count_index] or 0)
    return [key for key, _ in sorted(totals.items(), key=lambda item: (-item[1], item[0]))[:20]]


def _header(sheet, left, right, row, title):
    sheet.cell(row, left, title)
    sheet.cell(row, right, "數量")
    for column in range(left, right + 1):
        cell = sheet.cell(row, column)
        cell.fill = PatternFill("solid", fgColor=LIGHT_BLUE)
        cell.font = Font(bold=True)
    sheet.cell(row + 1, left, "依輸入資料自動加總").font = Font(italic=True, color="4B5563", size=10)


def _section(sheet, left, right, title, labels, source_letter, count_letter, last_row, start_row, total_row):
    _header(sheet, left, right, start_row - 2, title)
    for offset, label in enumerate(labels):
        row = start_row + offset
        sheet.cell(row, left, label)
        sheet.cell(row, right, f"=SUMIF('原始資料'!${source_letter}$2:${source_letter}${last_row},{get_column_letter(left)}{row},'原始資料'!${count_letter}$2:${count_letter}${last_row})")
        sheet.cell(row, right).fill = PatternFill("solid", fgColor=VALUE_BLUE)
        sheet.cell(row, right).number_format = "#,##0"
        for column in range(left, right + 1):
            sheet.cell(row, column).border = Border(bottom=THIN)
    sheet.cell(total_row, left, f"{title}合計")
    sheet.cell(total_row, right, f"=SUM('原始資料'!${count_letter}$2:${count_letter}${last_row})")
    for column in range(left, right + 1):
        cell = sheet.cell(total_row, column)
        cell.fill = PatternFill("solid", fgColor=LIGHT_BLUE)
        cell.font = Font(bold=True)
        cell.border = Border(bottom=Side(style="double", color="7F8C8D"))
    sheet.cell(total_row, right).number_format = "#,##0"


def build_workbook(data: dict, output_path: str | Path):
    headers, rows = list(data.get("headers", [])), list(data.get("rows", []))
    missing = [name for name in REQUIRED if name not in headers]
    if missing:
        raise ValueError(f"Pivot Cache 缺少必要欄位：{'、'.join(missing)}")
    index = {name: position for position, name in enumerate(headers)}
    count_index = index["件數"]
    normalized_rows = [list(row) + [""] * max(0, len(headers) - len(row)) for row in rows]
    last_row = len(normalized_rows) + 1
    field = {name: get_column_letter(position + 1) for name, position in index.items()}

    workbook = Workbook()
    source = workbook.active
    source.title = "原始資料"
    source.sheet_view.showGridLines = False
    source.append([*headers, "路口"])
    for row in normalized_rows:
        source.append(row + [None])
    intersection_column = get_column_letter(len(headers) + 1)
    for row in range(2, last_row + 1):
        source.cell(row, len(headers) + 1, f'=IF(AND(${field["路段"]}{row}<>"",${field["交叉路名"]}{row}<>""),${field["路段"]}{row}&"／"&${field["交叉路名"]}{row},"")')
    end_column = get_column_letter(len(headers) + 1)
    table = Table(displayName="RecoveredPivotData", ref=f"A1:{end_column}{last_row}")
    table.tableStyleInfo = TableStyleInfo(name="TableStyleMedium2", showRowStripes=False, showColumnStripes=False)
    source.add_table(table)
    for cell in source[1]:
        cell.fill = PatternFill("solid", fgColor=BLUE)
        cell.font = Font(bold=True, color="FFFFFF")
    source.freeze_panes = "A2"
    for column in range(1, len(headers) + 2):
        source.column_dimensions[get_column_letter(column)].width = 15
    source.column_dimensions[field["路段"]].width = 22
    source.column_dimensions[field["肇事原因"]].width = 34

    sheet = workbook.create_sheet("分析表")
    sheet.sheet_view.showGridLines = False
    sheet.merge_cells("A1:H1")
    sheet["A1"] = "交通事故統計分析"
    sheet["A1"].fill = PatternFill("solid", fgColor=BLUE)
    sheet["A1"].font = Font(bold=True, color="FFFFFF", size=14)
    sheet["A1"].alignment = Alignment(horizontal="center")
    sheet.row_dimensions[1].height = 28
    sheet.merge_cells("A2:H2")
    sheet["A2"] = "本表由樞紐快取自動還原；藍底數量為公式加總。"
    sheet["A2"].font = Font(italic=True, color="4B5563", size=10)

    roads = _rank(normalized_rows, index["路段"], count_index)
    intersections = _intersection_rank(normalized_rows, index["路段"], index["交叉路名"], count_index)
    periods = _rank(normalized_rows, index["發生時間"], count_index, 24)
    causes = _rank(normalized_rows, index["肇事原因"], count_index)
    vehicles = _rank(normalized_rows, index["當事者區分"], count_index)
    _section(sheet, 1, 2, "易肇事路段", roads, field["路段"], field["件數"], last_row, 6, 27)
    _section(sheet, 4, 5, "易肇事路口", intersections, intersection_column, field["件數"], last_row, 6, 27)
    _section(sheet, 7, 8, "易肇事時段", periods, field["發生時間"], field["件數"], last_row, 6, 31)
    _section(sheet, 1, 2, "易肇事肇因", causes, field["肇事原因"], field["件數"], last_row, 43, 64)
    _section(sheet, 7, 8, "易肇事車種", vehicles, field["當事者區分"], field["件數"], last_row, 43, 64)
    _header(sheet, 4, 5, 40, "易肇事年齡")
    age_labels = ["0-17", "18-20", "21-30", "31-40", "41-50", "51-60", "61-70", "71歲以上", "沒有年紀", "合計"]
    for offset, label in enumerate(age_labels, 42):
        sheet.cell(offset, 4, label)
    ranges = ((0, 17), (18, 20), (21, 30), (31, 40), (41, 50), (51, 60), (61, 70))
    for row, (low, high) in enumerate(ranges, 42):
        sheet.cell(row, 5, f'=SUMIFS(\'原始資料\'!${field["件數"]}$2:${field["件數"]}${last_row},\'原始資料\'!${field["年齡"]}$2:${field["年齡"]}${last_row},">={low}",\'原始資料\'!${field["年齡"]}$2:${field["年齡"]}${last_row},"<={high}")')
    sheet["E49"] = f'=SUMIFS(\'原始資料\'!${field["件數"]}$2:${field["件數"]}${last_row},\'原始資料\'!${field["年齡"]}$2:${field["年齡"]}${last_row},">=71")'
    sheet["E50"] = f'=SUM(\'原始資料\'!${field["件數"]}$2:${field["件數"]}${last_row})-SUM(E42:E49)'
    sheet["E51"] = "=SUM(E42:E50)"
    for row in range(42, 52):
        for column in range(4, 6):
            sheet.cell(row, column).border = Border(bottom=THIN)
        sheet.cell(row, 5).fill = PatternFill("solid", fgColor=VALUE_BLUE)
        sheet.cell(row, 5).number_format = "#,##0"
    for column in range(4, 6):
        cell = sheet.cell(51, column)
        cell.fill = PatternFill("solid", fgColor=LIGHT_BLUE)
        cell.font = Font(bold=True)
        cell.border = Border(bottom=Side(style="double", color="7F8C8D"))
    for column, width in {"A": 28, "B": 13, "C": 4, "D": 32, "E": 13, "F": 4, "G": 34, "H": 13}.items():
        sheet.column_dimensions[column].width = width
    sheet.freeze_panes = "A5"
    workbook.save(output_path)


def main():
    if len(sys.argv) != 3:
        raise SystemExit("用法：build_xlsx.py 輸入.json 輸出.xlsx")
    build_workbook(json.loads(Path(sys.argv[1]).read_text(encoding="utf-8")), sys.argv[2])


if __name__ == "__main__":
    main()
